source/common/grpc/async_client_impl.cc

```diff
@@ -122,24 +121,34 @@ void AsyncStreamImpl::onData(Buffer::Instance& data, bool end_stream) {
     }
     callbacks_.onReceiveMessageUntyped(std::move(response));
   }
+
+  if (end_stream) {
+    Http::HeaderMapPtr empty_trailers = std::make_unique<Http::HeaderMapImpl>();
+    callbacks_.onReceiveTrailingMetadata(std::move(empty_trailers));
+    streamError(Status::GrpcStatus::Unknown);
+    return;
```

Comment:
nit: return not needed

Trying to understand this change in general. So we might raise some data, but then fail the stream after that?

Reply:
Yeah; matching the behavior in `Grpc::GoogleAsyncClientImpl` here. It has to do with how we handle the situation of failure. The Google gRPC client basically just tells you something like "receives from now on aren't possible, so you must issue a Finish operation". You issue the Finish operation and it then gives you status and trailing metadata at some point in the future.

I'm not sure about the validity of trailing metadata on error scenarios like this, but I'm seeing the Google gRPC C++ client library deliver the message even in the above error scenario.

@vjpai for gRPC C++ team visibility.

Reply:
Sorry, I joined this conversation in the middle so I don't know the full-context, but it is our principle in gRPC that client-side `Finish` is _always_ valid. It's always valid to know the outcome of the RPC that you started, and any trailing metadata that accompanies the `Finish` is valid whether or not it's a `Status::OK` response. Was that what this topic was about?

Reply:
@vjpai I think what you write is basically what we've implemented, so it's all good. Thanks for the clarification.